vassal-app/src/main/java/bsh/BeanShellExpressionValidator.java

```diff
@@ -1,187 +1,213 @@
+ * package visible elements in the bsh library.
+ *
+ */
+public class BeanShellExpressionValidator {
+  
+  protected String expression;
+  protected List<String> variables = new ArrayList<>();
+  protected List<String> stringVariables = new ArrayList<>();
+  protected List<String> methods = new ArrayList<>();
+  protected String error;
+  protected boolean valid;
+
+  protected final String SUPPORTED_STRING_METHODS = ".length,.contains,.startsWith,.endsWith,.matches,.indexOf,.lastIndexOf,.substring,.replace,";
+
+  /**
+   * Build a new Validator and validate the expression
+   * @param expression Expression to validate
+   */
+  public BeanShellExpressionValidator(String expression) {
+    this.expression = expression; 
+    valid = validate();
+  }
+  
+  /**
+   * Is the expression valid?
+   * @return valid
+   */
+  public boolean isValid() {
+    return valid;
+  }
+  
+  protected void setValid(boolean b) {
+    valid = b;
+  }
+
+  /**
+   * Return a list of Variable references in the expression that we know must be Strings
+   * @return List of variables
+   */
+  public List<String> getStringVariables() {
+    return stringVariables;
+  }
+
+  /**
+   * Return a list of Variable references in the expression
+   * @return List of variables
+   */
+  public List<String> getVariables() {
+    return variables;
+  }
+  
+  /**
+   * Return a list of Methods called by the expression
+   * @return List of Methods
+   */
+  public List<String> getMethods() {
+    return methods;
+  }
+  
+  /**
+   * Return an Error Message if no valid
+   * @return Error message
+   */
+  public String getError() {
+    return error;
+  }
+  
+  protected void setError(String s) {
+    error = s;
+  }
+  
+  /**
+   * Validate the expression
+   * 
+   * @return Expression validity
+   */
+  protected boolean validate() {
+    final String expr = stripBraces(expression);
+    
+    setError("");
+    try {
+      Parser p = new Parser(new StringReader(expr + ";"));
+      for (;;) {
+        if (p.Line()) {
+          return true;
+        } 
+        else {
+          final SimpleNode node = p.popNode();
+          if (! processNode(node)) {
+            return false;
+          }
+        }
+      }
+    }
+    catch (ParseException e) {
+      setError(e.getMessage());
+      return false;
+    }
+    catch (TokenMgrError e) {
+      setError(e.getMessage());
+      return false;
+    }
+  }
+  
+  /**
+   * If the expression is surrounded by Vassal expression braces {}
+   * replace them with spaces so that it will validate and report errors
+   * in the correct location
+   * 
+   * @param s Expression
+   * @return stripped expression
+   */
+  public static String stripBraces(String s) {
+    String expr = s;
+    if (s.trim().startsWith("{") && s.trim().endsWith("}")) {
+      final int start = s.indexOf("{");
+      final int end = s.lastIndexOf("}");
+      StringBuilder buffer = new StringBuilder(s.length());
```

Comment:
I don't get this. If `s` starts with `{`, don't you already know that the index of `{` is 0, and if it ends with `}` that the index of the last `}` is `s.length()-1`?

Reply:
No, s.trim() starts with "{", not s. There may be extra leading or trailing spaces that I need to take into account, but it is not stripBraces() job to remove them. The purpose is to replace the outermost enclosing {}, if it exists with space characters, but not otherwise disrupt the the string.

Could probably do this better, but it is Module Editor only and limited use, I'm not too fussed about efficiency

Reply:
I misread. Thanks.